consensus/src/pow/work.rs

```diff
@@ -0,0 +1,133 @@
+#![allow(dead_code)]
+
+use crate::pow::helpers::{
+    calculate_new_target, check_difficulty, is_for_retarget, retarget_block_time, special_rules,
+};
+use crate::pow::temp::BlockIndex;
+use crate::pow::{Error, PoW};
+use common::chain::block::Block;
+use common::primitives::consensus_data::{ConsensusData, PoWData};
+use common::primitives::{Compact, Idable};
+use common::Uint256;
+
+pub fn check_proof_of_work(hash: Uint256, bits: Compact) -> bool {
+    if let Ok(target) = Uint256::try_from(bits) {
+        hash > target
+    } else {
+        false
+    }
+}
+
+impl PoW {
+    /// The difference (in block time) between the current block and 2016th block before the current one.
+    /// This difference should be inclusive between (2 weeks/4) and (2 weeks*4).
+    /// See Bitcoin's Protocol rules on [Difficulty change](https://en.bitcoin.it/wiki/Protocol_rules)
+    pub fn actual_timespan(&self, prev_block_blocktime: u32, retarget_blocktime: u32) -> u64 {
+        let actual_timespan = (prev_block_blocktime - retarget_blocktime) as u64;
+
+        // 2 weeks / 4
+        let lower_bound = self.min_target_timespan_in_secs();
+        // 2 wees * 4
+        let upper_bound = self.max_target_timespan_in_secs();
+
+        if actual_timespan < lower_bound {
+            lower_bound
+        } else if actual_timespan > upper_bound {
+            upper_bound
+        } else {
+            actual_timespan
+        }
+    }
+
+    pub fn get_work_required(
+        &self,
+        prev_block_index: &BlockIndex,
+        new_block_time: u32,
+    ) -> Result<Compact, Error> {
+        //TODO: check prev_block_index exists
+        let prev_block_bits = prev_block_index.data.bits();
+
+        if self.no_retargeting() {
+            return Ok(prev_block_bits);
+        }
+
+        let current_height = prev_block_index
+            .height
+            .checked_add(1)
+            .ok_or_else(|| Error::OutofBounds("max block height has been reached.".to_string()))?;
+        let adjustment_interval = self.difficulty_adjustment_interval();
+
+        if is_for_retarget(adjustment_interval, current_height) {
+            let retarget_block_time = retarget_block_time(adjustment_interval, prev_block_index);
```

Comment:
This doesn't seem to be correct. The block that you use to recalculate the new nBits is the first block in the 2016-block interval. Then, if that is longer than two weeks, the difficulty goes down. If it's less than two weeks, the difficulty goes up.

Reply:
I will rename the function to `get_starting_block_time`. 

The function is using the `prev_block_index` to call the `get_ancestor` method, based on the adjustment interval.